pkg/security/keychain.go

```diff
@@ -58,14 +67,176 @@ func SecKeyUpdate(connectionID string, username string, password string) *SecErr
+		}
+		writeErr := ioutil.WriteFile(GetPathToInsecureKeyring(), body, 0644)
+		if writeErr != nil {
+			return &SecError{errOpKeyring, writeErr, writeErr.Error()}
+		}
+		return nil
+	}
+
+	// else store it in system keyring
+	err := keyring.Set(service, uName, pass)
+	if err != nil {
+		return &SecError{errOpKeyring, err, err.Error()}
+	}
+	return nil
+}
+
+// GetSecretFromKeyring gets the secret from either the system keyring or our insecure keyring.
+func GetSecretFromKeyring(connectionID, uName string) (string, *SecError) {
+	service := connectionIDToService(connectionID)
+	if globals.UseInsecureKeyring {
+		secrets, readErr := readInsecureKeyring()
+		if readErr != nil {
+			return "", readErr
+		}
+		for _, secret := range secrets {
+			sameService := string(secret.Service) == service
+			sameUsername := string(secret.Username) == uName
+			matchingSecret := sameService && sameUsername
+			if matchingSecret {
+				return string(secret.Password), nil
+			}
+		}
+		err := errors.New("secret not found in keyring")
+		return "", &SecError{errOpKeyring, err, err.Error()}
+	}
+	// else get from system keyring
+	secret, err := keyring.Get(service, uName)
 	if err != nil {
 		return "", &SecError{errOpKeyring, err, err.Error()}
 	}
 	return secret, nil
 }
+
+// DeleteSecretFromKeyring deletes the secret from either the system keyring or our insecure keyring.
+func DeleteSecretFromKeyring(connectionID, uName string) *SecError {
+	service := connectionIDToService(connectionID)
+	if globals.UseInsecureKeyring {
+		secrets, readErr := readInsecureKeyring()
+		if readErr != nil {
+			return readErr
+		}
+		indexOfSecretToDelete := -1
+		for i, secret := range secrets {
+			sameService := string(secret.Service) == service
+			sameUsername := string(secret.Username) == uName
+			matchingSecret := sameService && sameUsername
+			if matchingSecret {
+				indexOfSecretToDelete = i
+			}
+		}
+		if indexOfSecretToDelete == -1 {
+			err := errors.New("secret not found in keyring")
+			return &SecError{errOpKeyring, err, err.Error()}
+		}
+		// remove existing secret
+		secrets = append(secrets[:indexOfSecretToDelete], secrets[indexOfSecretToDelete+1:]...)
+		if len(secrets) == 0 {
+			err := os.Remove(GetPathToInsecureKeyring())
+			if err != nil {
+				return &SecError{errOpKeyring, err, err.Error()}
+			}
+			return nil
+		}
+		body, marshallErr := json.MarshalIndent(secrets, "", "\t")
+		if marshallErr != nil {
+			return &SecError{errOpKeyring, marshallErr, marshallErr.Error()}
+		}
+		writeErr := ioutil.WriteFile(GetPathToInsecureKeyring(), body, 0644)
+		if writeErr != nil {
+			return &SecError{errOpKeyring, writeErr, writeErr.Error()}
+		}
+		return nil
+	}
+	// else delete from system keyring
+	err := keyring.Delete(service, uName)
+	if err != nil {
+		return &SecError{errOpKeyring, err, err.Error()}
+	}
+	return nil
+}
+
+func readInsecureKeyring() ([]KeyringSecret, *SecError) {
+	file, readErr := ioutil.ReadFile(GetPathToInsecureKeyring())
```

Comment:
What does `ioutil.ReadFile()` do if the file doesn't exist? Will that error?

Reply:
yea `readErr` is `open /Users/richard.waller@ibm.com/.codewind/config/insecureKeychain.json: no such file or directory`. I've just checked this and this automated test also verifies this: https://github.com/eclipse/codewind-installer/pull/405/files#diff-5692282167dc222ced4f1485c91cbad7R87. Whoever calls `readInsecureKeyring` will have to handle the err just like any other

Reply:
You make the same call when writing secrets, see: https://github.com/eclipse/codewind-installer/pull/405#discussion_r392221978 - can you write if the keychain file doesn't exist?

Reply:
yep https://github.com/eclipse/codewind-installer/pull/405#discussion_r392240264